### DISCUSSION TITLE
Delete Ast Element in Custom Action: How to determine the position until where to delete the text

### DISCUSSION THREAD

**Discussion by goto40** [eclipse-langium/langium#1337] [Q&A] [answered]
I use langium 2.1.3

* I have a custom action (like in https://github.com/eclipse-langium/langium/blob/main/packages/langium/src/grammar/lsp/grammar-code-actions.ts) to delete an unused part of my model.
* If I delete the text from range: `diagnostic.range.start .. diagnostic.range.end` I do not delete the model text as expected, because I want to delete the model text until the next model element starts.

**Not**
```
my-model-element1
my-model-element2-to-be-deleted // delete this with my action!
my-model-element1
```
--> 
```
my-model-element1
 // delete this with my action!
my-model-element1
```

**But**
```
my-model-element1
my-model-element2-to-be-deleted // delete this with my action!
my-model-element1
```
--> 
```
my-model-element1
my-model-element1
```

**How did I solve this**: I determine the "end position" to be deleted by calling `getNextNode(myModelElementToBeDeleted.$cstNode, false)?.range.start`

It works.

Is this the way it is intended to do this?

**Comment by msujew** [comment, marked as answer]
Hey @goto40,

since you want to delete all text from one AST node until the next node, it seems reasonable to use `getNextNode`. You shouldn't run into any issues with that such as unintentionally deleting more code than intended. We use `getNextNode` for some custom CST validations in our own projects quite regularly.

> If I delete the text from range: diagnostic.range.start .. diagnostic.range.end I do not delete the model text as expected, because I want to delete the model text until the next model element starts.

What is "expected" in this regard would be up for discussion. Since the diagnostic is only applied to the actual model text (but not the whole line), it will only delete what actually belongs to the model. As the comment does not belong to it, it will stay right where it is. I would say that this is the expected behavior for this functionality.

**Reply by goto40** [reply]
Thank you. The  comment is deleted in my case... I thought the "false" argument passed to `getNextNode` was ment for that...
